src/merge.ts

```diff
@@ -0,0 +1,25 @@
+import { EventPayload } from './EventPayload';
+
+/** Used for merging request JSON into an existing EventPayload. Returns a new EventPayload object */
+export const merge = (json1: EventPayload, json2: Record<string, unknown>): EventPayload => {
+    const copy: Record<string, unknown> = { ...json1 };
+
+    for (const key in json2) {
+        if (json2.hasOwnProperty(key)) {
+            const value = json2[key];
+            if (value === null || value === undefined) {
+                delete copy[key];
+            } else if (typeof value !== 'object') {
+                copy[key] = value;
+            } else {
+                copy[key] = merge(copy[key] as EventPayload, value as Record<string, unknown>);
+            }
+        }
+    }
+
+    const result: EventPayload = {
+        ...copy,
+        action: copy.action as EventPayload['action']
```

Comment:
Why do you need to do this? Does `const result: EventPayload = copy` not work?

Reply:
Unfortunately it does not. Doing that or 
 `const result: EventPayload = {...copy};`
results in 
`Property 'action' is missing in type '{ [x: string]: unknown; }' but required in type 'EventPayload'.ts(2741)`

Reply:
Gotcha. Also, isn't the spread operator `...` result in this being a shallow copy?

Reply:
It makes a deep copy at the top level, but if it has nested objects then the children are shallow copies: https://www.educative.io/answers/what-is-the-spread-operator-in-javascript

However this is not an issue because we do not ever change the original.

Reply:
Appears I was wrong about this: [See example](https://www.typescriptlang.org/play?#code/FAYw9gdgzgLgBAWwKYCcDmS4F44AowoCWahEAhgDYBccAogG5IQwAKZAnhWGQCYA0cCEgDuANUoBXJFBpsUMQpTaduPAJQ0GTVhy69sAPjgBvYHDjho8cAAd22E3AB0LgsVKU4AXwDcZi5CwcLBkIADWNACCKCgcADwA2nIKSrqqAsmKFMp6PAC6RjgJCbbsAkJiktJ5eX7+wgAWhBSYuCHhTi0QaDANaib+5pZBCTBk6EgwAlBgEiggSHkO7WFONmA2uP0A-NtwCbWDcIQAZnhjE-AAZFfBs-NI-abmL3AA8gBGAFZIIDBOYSQ7CgbXuCzUThOBFooQauFwgPY-SwRmerxew3gWmYAGkgQ5EXAyFA4IiwGdsToVLw-OiMYF4PQqsswUgEpS8exDnTzKc8EyKFJsFgcBAJBQKHAAD5SuACoUinASCA8JAnUhIdQDHkvVUtGCYC4YGDsxi4oHcnleOBIChQTB83Awdg2JDkuXMgCEIrgAHIwN9fjBfU8jnSjZNTdpOUscPKkLSrTa7Zg0TrHZ6IyaORbQzrXlmo+auQ5jL4w+ivBXXis1hIoHDRuNjUWYDGiSTMqlqfwPYLFmpE3Sq8PB0cR95gP4UJM5hBtS8XE5SnwjqEFJAaKUnOvCJAO3QzVTcglfbvIL68v5y+XQAy4G4SORJThjDANjRjEJYDQAERQsBfy8KsAHoQLgAABGAoAAWmICACCQYBkAmfAiCfSgBDfD8TG-GA-w+cYgK8NQp2GMAWk6MA0DQ9xnycd8bCcPD+jAuBf0IlBfynUgDRQE5QkwSkclUBcQIAKnEuAABUGkwetUCJP493nRpCBABoLDIBSeDgXpMCQI8BCQJw0CcOBIgAEUsgB9aS3hsgBhSIACVpIfFAZIACQAVQAWQAIQAZRsnyWHMcSQP8c8IBoWAiG6RMJKknyIEIABHIVCFVZhTkIRT3z0-K9OoyY5M8wztBJUgiWCUg0BaOAPhQMBhCgergmkdr90i-x7SgbqIAASR4bY4pgBK0CSySZLkuAfJcgAZB8zn0uAbDIDA4EaVBDTmyrmAfEAQDmGddN68wNowHyUAoMbggm+rpqk2TMAW5b3TWq69oMo9jhJHhCBnZTuj0ub6EIdr30899zIuuBVVgDwNwgG67vGybntmt6lpWsHMG+7amg0-G5Uhwhoa05A4BOFqEHWogCBK0mDv+OB4ZnE5UBnFA0fu+Knv8ZKLLgCgyA+W0O3atAhF0wq1tZ4zTPMsg4Ec6TIlF8XbThqLzDFiX0YezGhZm17RbAEBKEwT65oUzzGjAOAMCEWIDTl-aj11-wuCtlp+cexLTZeuaFGQEIEBsIl4DUkmFb+y2TpiTUBBqoagreGmCAQMg2fhsPpDGSP7ss3PMFlAWg-MYWAHU5P0mHPe0f7SZnKBxXgd0PjAeBdwh51vfMbuYADk3q5moaIChFAc5Rolu4keA1ohqGCABwz1MwMgVSalq2tQQfd9a+0UEiF2R8-I5hfN5rj8U4kZhARR3e2inNPj7RD6HveT9HwXXmvnNW++9PKMBQINKWltn6alfr0FmXt2Z6xeMAk+ohUCDT-lXF4gCbaujdh1KAwIDT0wflAsuulhBv3gZ-RBRwwBQEwVNK+Zs5obFQLnAhRCkD0zARA0hT9yGwPfk3ZgX8HxQDQeAlSjChxwBwQjDeCxIECJfpQuBH9RG0NeKqCGCwZHMJDpgHRm8LBiwGso6BFCqEaLzkg8wxiFiOTMQwjG-9sEsPkifIk58LGCLUcI36ND4bmHtmfbQ+jzDlnHuJQYUkABiTM1oAE0kAAA9rAUHyodIKlkcTVQgPQCijBdI1RsQIXokMW7bwfIGP4iAyA2BsPVWJpNyDhzxr0ehmAcl5OZsvdBKktaGxJLbLpcAelQCcLEpBIBMnaCCjwMI90XK-AIDwOIlc0DTEDmgAwWNJ6AytgaEkO0G7UMOpUtabcO5414qgZ8SkFD93YIfO5KBnw+V-jQbuFEkDb2ei0uuTBW6rJQLpduCAc5EAAF7SEQB3QgNhGqsygOUuaYoEAS08u6FFILwBgrgDOGwbcqpTJeIYm0f1LkzkEcSY4ZwKYtx4BIJF6lBEVJJDnCA9gC6TOmf4cAyoL6CAkJi1AALyXzSjoVAAjAABnWmQIGnLc4aQ6idWAYB6abNJECEkhUdXxn1U7fh0ChHnP+C0zkJJxiYCtvaOC0AmBQ0IIwHwfYpA2pnFpB19poAU1dSZflQx6zvgQNJTaLi4ArPxeszZ2zJp7ODi0ny0qnbysVcq+pMA1Wgw1WG42HVETGpFWK0BVQS2mrLuamxZKIpwGtUSb19qkCOr9S6xgdatEWFDVq8Q-Yo0xrWRsnZ5RRVYqTdE7GcAUnpJtLlZ0zNY4BMpc3fFbd1gqkmRFJB2gKbsDGkcWUaZ3EUoQIvcWLQBD5q1agGC9oYBjA+Mihd9ghqWSzo3Ayr7fGqOsSI2x1Z50KGdCNVxWDXgTnMMeis8jCCQovc+79IG30fofU1Hlc1CGwG4eZWSlTKnVLeQ899YjXh7udD5bKNAMVYtkZErG5shrhV4DwNuJJp4WqcPDRF90T3yOY-QAALESHgbGuq-pgf4rj3bzCsfY+BphACPFEgoGgNwvR6aFQUszbekB2AIEILC0mzHRPifMWQE4fEAgSiDCpMRlB1NEE0xE7wEr62T2nrPQZOldMDTIQaIkz52AKBACcqhqsNryHUuKcYcB7ZdpiZKiCM4oVhCgFavVTbMArwpkzZAnSeA2p3ka7LZNV6eXfXyyVnJFWPvuTQAABgAPQSGQGC0LIgwQAFpypggAThsjBPIxgZV8BlQANi8AAEka4luA-ahQbXq+8prrWAA6qSABMfXNsAHZaAjbGxNoTM3GuxJaRBGwEhn3qWDeVvLKBlmgrjaOwt3Qk1VmAM6V0cAuzZDSPoHA-24hDrBXEMkFIjwiV4AIZUYQELCAgAYPZQA)

Updating to make a deep copy